### PR TITLE
Make links in Web Apps more accessible

### PR DESCRIPTION
**Pull request by solleks**
## Summary

https://dimagi-dev.atlassian.net/browse/USH-394

Make links in Web Apps more accessible.

This is the first change to improve Web Apps accessibility based on feedback from screen reader users. The overall plan for this work is described in [this document](https://docs.google.com/document/d/1fHNaVlaFl1xrbtPbMqj0YpByss1e0_hzCRlAUbVLboM/edit?usp=sharing), which includes links to the feedback from a vendor and links to accessibility standards and resources.

This first change makes navigation links in Web Apps implement the [WAI-ARIA](https://www.w3.org/TR/wai-aria/) role of "link", which typically involves 4 parts:

1. find the appropriate HTML element and add the role="link" attribute which informs screen readers to treat it as a link.
2. put the element into the tab order with the tabindex="0" attribute, so it can receive keyboard focus.
3. add a keyboard event handler to the Marionette view for the element which treats Enter key presses like mouse clicks.
4. add a label with either aria-labelledby or aria-label to tell a screen reader what text to read for the element.

Using aria-labelledby, which refers to another element by id, is preferable to aria-label, so I tried to use aria-labelledby in most cases. Using aria-labelledby with translated terms should read them after translation, which definitely seems better.

## Product Description

For users without a screen reader, this change enables keyboard navigation of Web Apps, using the Tab and Enter keys. Navigating by keyboard is one way to test that it works. No other changes should be evident without a screen reader.

### QA Plan

The QA plan will be check navigation of these pages by mouse to ensure there are no regressions and to test navigation by keyboard to see if the changes actually work.

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-2188)

### Safety story

WAI-ARIA attributes are designed to have no effect outside of accessibility tools, so they shouldn't cause any regressions.

The other changes are putting elements into the focus ring and handling key clicks on them, which should not affect existing users who navigate by mouse clicks.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/users/views.js [hunk 1]
```diff
@@ -41,6 +41,14 @@ hqDefine("cloudcare/js/formplayer/users/views", function () {
         tagName: 'tr',
         events: {
             'click': 'onClickUser',
+            'keydown': 'onKeyActionUser',
+        },
+        attributes: function () {
+            return {
+                "role": "link",
+                "tabindex": "0",
+                "aria-label": this.model.get('username')
```
[the remark this answers is not in the document]
Comment (solleks) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/users/views.js]: Done

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/views.js [hunk 2]
```diff
@@ -9,13 +9,21 @@ hqDefine("cloudcare/js/formplayer/apps/views", function () {
...
             "click": "rowClick",
+            "keydown": "rowKeyAction",
         },
 
         rowClick: function (e) {
             e.preventDefault();
             FormplayerFrontend.trigger("app:select", this.model.get('_id'));
         },
 
+        rowKeyAction: function(e) {
```
[the remark this answers is not in the document]
Comment (solleks) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/views.js]: Done

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/views.js [hunk 3]
```diff
@@ -9,13 +9,21 @@ hqDefine("cloudcare/js/formplayer/apps/views", function () {
...
+            if (e.keyCode == 13) {
```
[the remark this answers is not in the document]
Comment (solleks) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/views.js]: Done

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/views.js [hunk 4]
```diff
@@ -48,6 +60,26 @@ hqDefine("cloudcare/js/formplayer/apps/views", function () {
             e.preventDefault();
             FormplayerFrontend.trigger("settings:list");
         },
+        incompleteSessionsKeyAction: function (e) {
+            if (e.keyCode == 13) {
```
[the remark this answers is not in the document]
Comment (solleks) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/views.js]: Done

**Review by stephherbers — COMMENTED** [review]
I got the chance to see this in action and the code itself looks good to me. I'm wondering if @orangejenny or anyone else more familiar with this part of the codebase has suggestions for the new additions during this first iteration of accessibility changes.

File: corehq/apps/cloudcare/templates/formplayer/session_list.html [hunk 5]
```diff
@@ -42,12 +42,12 @@ <h1 class="page-title">{% trans "Incomplete Forms" %}</h1>
...
   </td>
   <td class="module-column module-column-name module-column-name-session">
-    <h3>
+    <h3 id="<%- labelId %>">
       <b><%- title %></b>
       <% if (caseName) { %>(<%- caseName %>)<% } %>
       opened <%- humanDateOpened %>
     </h3>
-    <div class="module-delete-control">
+    <div class="module-delete-control" role="button" tabindex="0" aria-label="Delete form">
```
Comment (orangejenny) [inline corehq/apps/cloudcare/templates/formplayer/session_list.html:50]: I think this should be translated, with `trans_html_attr`

**Review by orangejenny — APPROVED** [review]
This looks great, thank you for writing out a thorough description.

Should the breadcrumb views be updated in this PR? I'd think they need similar changes, since they use "links" that are just `<li>` elements, not anchor tags.

Similarly, should `QueryView` be updated in this PR, or will it be in a future PR?

**Comment by solleks** [comment]
Thanks for the comments.

Kiran also pointed out the breadcrumbs, so we have [an issue](https://dimagi-dev.atlassian.net/browse/QA-2203) for that and I'll try to resolve it in this PR. I will use trans_html_attr as well.

Is QueryView related to Data Preview? If so, I have not really looked at Data Preview yet, since it wasn't mentioned in the accessibility review, but Kiran called it out in [this issue](https://dimagi-dev.atlassian.net/browse/QA-2204). If it's something different, how can I create a Web Apps page that uses QueryView?

**Comment by orangejenny** [comment]
Query View, also called case claim or case search, is different than Data Preview. It's a feature that lets users search HQ for cases and then pull those cases down onto their phone. There are a couple of steps to setting it up, I'll send you an email.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 6]
```diff
@@ -422,15 +461,27 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
...
             "click": "crumbClick",
+            "keydown": "crumbKeyAction",
         },
 
         crumbClick: function (e) {
             e.preventDefault();
             var crumbId = this.options.model.get('id');
             FormplayerFrontend.trigger("breadcrumbSelect", crumbId);
         },
+        crumbKeyAction: function(e) {
```
[the remark this answers is not in the document]
Comment (solleks) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js]: Fixed.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 7]
```diff
@@ -422,15 +461,27 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
         tagName: "li",
         template: _.template($("#breadcrumb-item-template").html() || ""),
         className: "breadcrumb-text",
+        attributes: function() {
```
[the remark this answers is not in the document]
Comment (solleks) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js]: Fixed.

File: corehq/apps/cloudcare/templates/formplayer/session_list.html (hunk 5 above)
Comment (solleks) [reply corehq/apps/cloudcare/templates/formplayer/session_list.html]: Done.

**Comment by solleks** [comment]
I heard there is pressure to ship some accessibility fixes, so I've tried to simplify this PR.

Breadcrumb fixes are in. Pagination fixes are out due to a merge conflict. I plan to leave Data Preview, Remote Case Search & Claim (QueryView) and other features noted in QA testing to future work.

**Comment by solleks** [comment]
Hi @stephherbers, if you want to double-check my work and merge it as you suggested, I think it is ready. But feel free to request more changes or more reviewers if I've added too much since @orangejenny's approval.

**Review by orangejenny — APPROVED** [review]
New commits look good to me.

**Pull request merged**